runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -10495,6 +10495,306 @@ inlineCompareAndSwapNative(
+   generateRegImmInstruction(TR::InstOpCode::AND4RegImm4, node, loopLimitReg, -16, cg);
+   generateRegRegInstruction(TR::InstOpCode::ADD4RegReg, node, loopLimitReg, offsetReg, cg);
+
+
+   // loopLabel
+   generateLabelInstruction(TR::InstOpCode::label, node, loopLabel, cg);
+
+   // if index >= loopLimit, jump to handling the residual bytes
+   generateRegRegInstruction(TR::InstOpCode::CMP8RegReg, node, indexReg, loopLimitReg, cg);
+   generateLabelInstruction(TR::InstOpCode::JGE1, node, residualLabel, cg);
+
+   // Load 16 bytes from address [buf + index]
+   generateRegMemInstruction(TR::InstOpCode::MOVDQURegMem, node, xmmChunkReg, generateX86MemoryReference(bufReg, indexReg, 0, TR::Compiler->om.contiguousArrayHeaderSizeInBytes(), cg), cg);
+
+   // Create zero vector
+   generateRegRegInstruction(TR::InstOpCode::PXORRegReg, node, xmmMaskReg, xmmMaskReg, cg);
+
+   // Compare 0 > chunk bytewise
+   generateRegRegInstruction(TR::InstOpCode::PCMPGTBRegReg, node, xmmMaskReg, xmmChunkReg, cg);
+
+   // Extract bitmask of negative values
+   generateRegRegInstruction(TR::InstOpCode::PMOVMSKB4RegReg, node, maskReg, xmmMaskReg, cg);
+
+   // Check if any negative values exist
+   generateRegRegInstruction(TR::InstOpCode::TEST4RegReg, node, maskReg, maskReg, cg);
+
+   // If the result is nonzero, handle finding a negative byte
+   if (recognizedMethod == TR::java_lang_StringCoding_hasNegatives)
+      {
+      // Jump to returnHasNegativesLabel
+      generateLabelInstruction(TR::InstOpCode::JNE4, node, returnHasNegativesLabel, cg);
+      }
+   else
+      {
+      // Break and calculate index of first negative byte
+      generateLabelInstruction(TR::InstOpCode::JNE1, node, foundNegativeLabel, cg);
+      }
+
+   // index += 16
+   generateRegImmInstruction(TR::InstOpCode::ADD8RegImm4, node, indexReg, 16, cg);
+
+   // Jump back to loopLabel
+   generateLabelInstruction(TR::InstOpCode::JMP1, node, loopLabel, cg);
+
+
+   // If we found any negative bytes, calculate the index of the first negative byte in the chunk
+   if (recognizedMethod == TR::java_lang_StringCoding_countPositives)
+      {
+      generateLabelInstruction(TR::InstOpCode::label, node, foundNegativeLabel, cg);
+      generateRegRegInstruction(TR::InstOpCode::TZCNT4RegReg, node, maskReg, maskReg, cg);
+      generateRegRegInstruction(TR::InstOpCode::ADD4RegReg, node, indexReg, maskReg, cg);
+
+      // Jump to returnHasNegativesLabel
+      generateLabelInstruction(TR::InstOpCode::JMP4, node, returnHasNegativesLabel, cg);
+      }
+
+
+   // Deal with the residual (last 15 or fewer) bytes
+   generateLabelInstruction(TR::InstOpCode::label, node, residualLabel, cg);
+
+   // bytesLeft = limit - loopLimit
+   generateRegRegInstruction(TR::InstOpCode::MOV8RegReg, node, bytesLeftReg, limitReg, cg);
+   generateRegRegInstruction(TR::InstOpCode::SUB8RegReg, node, bytesLeftReg, loopLimitReg, cg);
```

Comment:
`loopLImitReg` isn't used after this point.  I wonder whether it could be used in place of `bytesLeftReg`, like this, to reduce the number of registers needed:

```suggestion
   // Calculate bytes remaining:  loopLimit = -(loopLimit - limit)
   generateRegRegInstruction(TR::InstOpCode::SUB4RegReg, node, loopLimitReg, limitReg, cg);
   generateRegInstruction(TR::InstOpCode::NEG4Reg, node, loopLimitReg, cg);
```